core/src/consensus/fork_choice.rs

```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
+        resetting to slot {:?} for now,
+        switch proof stake: {},
+        threshold stake: {},
+        total stake: {}",
+        heaviest_bank_slot,
+        reset_bank.as_ref().map(|b| b.slot()),
+        switch_proof_stake,
+        total_stake as f64 * SWITCH_FORK_THRESHOLD,
+        total_stake
+    );
+    failure_reasons.push(HeaviestForkFailures::FailedSwitchThreshold(
+        heaviest_bank_slot,
+        switch_proof_stake,
+        total_stake,
+    ));
+    switch_fork_decision
+}
+
+fn select_candidate_vote_and_reset_banks<'a>(
+    heaviest_bank: &'a Arc<Bank>,
+    heaviest_bank_on_same_voted_fork: Option<&'a Arc<Bank>>,
+    progress: &'a ProgressMap,
+    tower: &'a Tower,
+    failure_reasons: &mut Vec<HeaviestForkFailures>,
+    initial_switch_fork_decision: SwitchForkDecision,
+) -> CandidateVoteAndResetBanks<'a> {
+    match initial_switch_fork_decision {
+        SwitchForkDecision::FailedSwitchThreshold(switch_proof_stake, total_stake) => {
+            // Even though we think we want to switch to a new, heavier fork, we
+            // actually want to continue voting on our current fork if our last
+            // vote was unable land.
+            let final_switch_fork_decision = select_forks_failed_switch_threshold(
+                heaviest_bank_on_same_voted_fork.map(|bank| bank.as_ref()),
+                progress,
+                tower,
+                heaviest_bank.slot(),
+                failure_reasons,
+                switch_proof_stake,
+                total_stake,
+                initial_switch_fork_decision,
+            );
+            let candidate_vote_bank = if final_switch_fork_decision.can_vote() {
+                heaviest_bank_on_same_voted_fork
+            } else {
+                Some(heaviest_bank)
+            };
+            CandidateVoteAndResetBanks {
+                candidate_vote_bank,
+                reset_bank: heaviest_bank_on_same_voted_fork,
+                switch_fork_decision: final_switch_fork_decision,
+            }
+        }
+        SwitchForkDecision::FailedSwitchDuplicateRollback(latest_duplicate_ancestor) => {
+            let reset_bank = Some(heaviest_bank);
+            info!(
+                "Waiting to switch vote to {}, resetting to slot {:?} for now, latest duplicate ancestor: {:?}",
+                heaviest_bank.slot(),
+                reset_bank.as_ref().map(|b| b.slot()),
+                latest_duplicate_ancestor,
+            );
+            failure_reasons.push(HeaviestForkFailures::FailedSwitchThreshold(
+                heaviest_bank.slot(),
+                0, // In this case we never actually performed the switch check, 0 for now
+                0,
+            ));
+            CandidateVoteAndResetBanks {
+                candidate_vote_bank: None,
+                reset_bank,
+                switch_fork_decision: initial_switch_fork_decision,
+            }
+        }
+        _ => CandidateVoteAndResetBanks {
```

Comment:
I know you just copy pasted this but I think we should use:
```rust
SwitchForkDecision::SameFork | SwitchForkDecision::SwitchProof(_) => ...
```
and avoid the wildcard